### PR TITLE
Justin/detail screens - The Gym Detail Screen

### PR DESCRIPTION
**Pull request by thisjustin123**
This PR includes the entire Gym Detail Screen. All composables are designed to take in and display information for arbitrary gyms. All screenshots are taken to display an example gym I made as a placeholder, as networking is not yet tackled.

**Changes:**
- Gym Header with name and "CLOSED" status (if closed based off system time)
- Expandable hours tab to see hours for today/the week.
- An interactable Popular Times widget to select times and see gym availability at certain times of the day.
- A facilities tab that shows Equipment and Miscellaneous info, as well as Gymnasium, Swimming, and Bowling info for each day of the week.
- A classes list for classes taking place that day. 
_(Note: In the demo gym, I filled this data with classes not necessarily taking place on today, but that is just for testing and would be sorted when actual information is pulled from backend.)_

**Notes:**

- Many clean animations :-) Please watch the demo video.

**Screenshots:**
Screenshot_20230305-180152_1_20 Screenshot_20230305-180212_1_20

**Animation Demo:**

https://user-images.githubusercontent.com/47724806/222991345-a9768816-2584-4e8c-b000-6f1f133e6728.mp4

### CODE REVIEW

File: app/src/main/java/com/cornellappdev/uplift/models/TimeModels.kt [hunk 1]
```diff
@@ -8,19 +8,87 @@ data class TimeInterval(
...
-    val minute: Int,
-    val isAM: Boolean
+    /** A minute between 0 and 59, inclusive. */
+    val minute: Int = 0,
+    val isAM: Boolean = true
 ) {
+    /**
+     * Returns a new [TimeOfDay] created by advancing the current time of day by [deltaHours] and [deltaMinutes].
+     */
+    fun getTimeLater(deltaMinutes: Int, deltaHours: Int): TimeOfDay {
```
Comment (Corfish123) [inline app/src/main/java/com/cornellappdev/uplift/models/TimeModels.kt:48]: what is the purpose of this function?

File: app/src/main/java/com/cornellappdev/uplift/ui/components/DayOfWeekSelector.kt [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+                    selectedDay = i
+                    onDaySelected(i)
+                }
+            ) {
+                Surface(
+                    modifier = Modifier
+                        .size(24.dp)
+                        .scale(animatedScale.value),
+                    shape = CircleShape,
+                    color = if (selectedDay == i) PRIMARY_YELLOW else if (today == i) GRAY01 else PRIMARY_YELLOW
```
Comment (Corfish123) [inline app/src/main/java/com/cornellappdev/uplift/ui/components/DayOfWeekSelector.kt:58]: shouldn't it be if (today == i) then have yellow not gray?

File: app/src/main/java/com/cornellappdev/uplift/ui/components/gymdetail/GymFacilitySection.kt [hunk 3]
```diff
@@ -0,0 +1,242 @@
...
+                text = title,
+                fontFamily = montserratFamily,
+                fontSize = 16.sp,
+                fontWeight = FontWeight(500),
+                lineHeight = 19.5.sp,
+                textAlign = TextAlign.Left,
+                modifier = Modifier.padding(start = 8.dp)
+            )
+            Spacer(modifier = Modifier.weight(1f))
+            if (open != OpenType.NOT_APPLICABLE)
```
Comment (Corfish123) [inline app/src/main/java/com/cornellappdev/uplift/ui/components/gymdetail/GymFacilitySection.kt:219]: what happens when it is not applicable? is it just blank?

Comment (Corfish123) [reply app/src/main/java/com/cornellappdev/uplift/ui/components/gymdetail/GymFacilitySection.kt:219]: nvm i see it just doesn't appear

File: app/src/main/java/com/cornellappdev/uplift/ui/components/gymdetail/GymSwimmingSection.kt [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+import com.cornellappdev.uplift.util.LIGHT_YELLOW
+import com.cornellappdev.uplift.util.PRIMARY_YELLOW
+import com.cornellappdev.uplift.util.montserratFamily
+
+/**
+ * Displays swimming information for the given gym throughout the week, defaulting to the day
+ * given by parameter [today].
+ */
+@Composable
+fun GymSwimmingSection(today: Int, gym: Gym) {
```
Comment (Corfish123) [inline app/src/main/java/com/cornellappdev/uplift/ui/components/gymdetail/GymSwimmingSection.kt:24]: general question, is there normally multiple times in a day for a gym.

File: app/src/main/java/com/cornellappdev/uplift/ui/components/gymdetail/PopularTimes.kt [hunk 5]
```diff
@@ -0,0 +1,201 @@
...
+                        ) {
+                            if (selectedPopularTime == i) deselect()
+                            else {
+                                selectedPopularTime = i
+                                lastSelectedPopularTime = i
+                            }
+                        }
+                ) {
+                    // The yellow bar.
+                    Surface(
```
Comment (Corfish123) [inline app/src/main/java/com/cornellappdev/uplift/ui/components/gymdetail/PopularTimes.kt:160]: I can't believe you did this without using a library. Well done

**Review by Corfish123 — APPROVED** [review]
Overall it is really impressive. Well done and keep up the good work.

File: app/src/main/java/com/cornellappdev/uplift/models/TimeModels.kt (hunk 1 above)
Comment (thisjustin123) [reply app/src/main/java/com/cornellappdev/uplift/models/TimeModels.kt:48]: I believe I used it to make implementing PopularTimes.kt easier. Basically how I designed PopularTimes is to just have a starting TimeOfDay, and then each time after is basically just some # of hours added onto the start TimeOfDay. This function thus allows me to easily represent every TimeOfDay that the PopularTimes must represent while only needing to pass in one particular TimeOfDay.

File: app/src/main/java/com/cornellappdev/uplift/ui/components/DayOfWeekSelector.kt (hunk 2 above)
Comment (thisjustin123) [reply app/src/main/java/com/cornellappdev/uplift/ui/components/DayOfWeekSelector.kt:58]: Oh right, there's a distinction here between a "selected" day and "today." Basically, no matter what, if the user selects a day, that day should be yellow. If a day is not selected but it is actually TODAY (as in today's day of the week), that day will appear with a greyed icon to remind the user which day is today if they'd like to go back to it.

File: app/src/main/java/com/cornellappdev/uplift/ui/components/gymdetail/GymSwimmingSection.kt (hunk 4 above)
Comment (thisjustin123) [reply app/src/main/java/com/cornellappdev/uplift/ui/components/gymdetail/GymSwimmingSection.kt:24]: Not generally, no, but I designed everything with as much flexibility as possible to be able to handle (hopefully) all cases. I guess it gets a bit messy in some places, but I hope my documentation accounts for that.

File: app/src/main/java/com/cornellappdev/uplift/MainActivity.kt [hunk 6]
```diff
@@ -3,14 +3,21 @@ package com.cornellappdev.uplift
...
+import com.cornellappdev.uplift.ui.viewmodels.GymDetailViewModel
+import com.cornellappdev.uplift.util.testMorrison
 
 class MainActivity : ComponentActivity() {
     override fun onCreate(savedInstanceState: Bundle?) {
         super.onCreate(savedInstanceState)
+        val gymDetailViewModel = GymDetailViewModel()
+
+        gymDetailViewModel.selectGym(testMorrison)
+
```
Comment (ckdesir) [inline app/src/main/java/com/cornellappdev/uplift/MainActivity.kt:17]: Sorry small comment, viewmodels should not be created like this but should be referenced using viewModel(), see[ this link](https://developer.android.com/jetpack/compose/libraries#viewmodel) for an example.

viewModel() returns an existing ViewModel or creates a new one in the given scope. The ViewModel is retained as long as the scope is alive. For example, if the composable is used in an activity, viewModel() returns the same instance until the activity is finished or the process is killed.

Comment (thisjustin123) [reply app/src/main/java/com/cornellappdev/uplift/MainActivity.kt:17]: Oh woah, I had no idea there was a sort of official support for viewmodels... thanks for showing me this! I will amend this in my next PR.

**Pull request merged**